Range Locking - rebased on latest fb-mysql-8.0.20

spetrunia opened the pull request:
This includes:
1. Iterator patches, re-based on top of the latest fb-mysql-8.0.20 tree 
2. Range Locking patch

spetrunia committed:
This adds a global my.cnf parameter, rocksdb_use_range_locking.

(Cherry-picked on top of the pushed iterator patches)

When it is ON, MyRocks will:
- initialize RocksDB to use range-locking lock manager
- for all DML operations (including SELECT .. FOR UPDATE) will lock
the scanned range before reading/modifying rows.
- In range locking mode, there is no snapshot checking (cannot do that
  for ranges). Instead, MyRocks will read and modify latest committed
  data, just like InnoDB does (in the code, grep for (start|end)
  _ignore_snapshot)
- Queries that do not have a finite range to scan, like
  UPDATE t1 .... ORDER BY t1.key LIMIT n
  will use a  "Locking iterator" which will read rows, lock the range,
  and re-read the rows. See class LockingIterator.

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    while (1) {
+      /*
```
Could you thd->killed in this loop and terminate/return HA_ERR_QUERY_INTERRUPTED if set?

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,190 @@
+//    lock_iter->Next();
+//    lock_iter->Valid()==true && lock_iter->key() == 'efg'
+//
+//   Now, the range ['bcd'.. 'efg'] (bounds incluive) is also locked, and there are no
```
nit: inclusive

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,190 @@
+//
+class LockingIterator : public rocksdb::Iterator {
+
+  rocksdb::Transaction *txn_;
```
Could these member names be prefixed with m_?

hermanlee reviewed (commented):
Just a few initial comments. I will continue to read through this.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -14738,6 +14641,36 @@ static int show_rocksdb_stall_vars(THD *thd MY_ATTRIBUTE((unused)),
+    DEF_STATUS_VAR_FUNC("current_lock_memory",
+                        &rocksdb_locktree_current_lock_memory, SHOW_LONGLONG),
+    // end of the array marker
+    {NullS, NullS, SHOW_LONG}};
```
Should  {NullS, NullS, SHOW_LONG, SHOW_SCOPE_GLOBAL}

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -14738,6 +14641,36 @@ static int show_rocksdb_stall_vars(THD *thd MY_ATTRIBUTE((unused)),
+    {NullS, NullS, SHOW_LONG}};
+
+static SHOW_VAR rocksdb_empty_status_variables[] = {
+    {NullS, NullS, SHOW_LONG}};
```
Should {NullS, NullS, SHOW_LONG, SHOW_SCOPE_GLOBAL}

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
-      if (rc != HA_EXIT_SUCCESS) {
-        return rc;
-      }
+  if (m_lock_rows == RDB_LOCK_NONE || !rocksdb_use_range_locking) {
```
For read-committed isolation level, we should probably skip range locks? It seems InnoDB read-committed isolation level doesn't hold range locks (except in short instances like inserting a new row).

spetrunia replied:
Yes, done.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+      end_slice= slice;
+    }
+  }
+  else if (find_flag == HA_READ_PREFIX_LAST) {
```
It seems this case can be combined with HA_READ_KEY_EXACT since both are setting the range lock at [(key,0) (key, +inf)]

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+  else if (find_flag == HA_READ_PREFIX_LAST) {
+    if (slice.size() == Rdb_key_def::INDEX_ID_SIZE) {
+      /* Reverse-ordered full index scan */
+      start_has_inf_suffix= true;
```
Even though start_has_inf_suffix is not used in the case of big_range = true, it's not clear why start_has_inf_suffix is true here? Isn't the range lock still [(INDEX_ID, 0), (INDEX_ID, +inf)], same as read_key_exact?

spetrunia replied:
Right. It should be false, but also it doesn't matter because it is not used.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+      slice= rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf),
+                            end_slice_size);
+
+      end_has_inf_suffix= false;
```
It seems these two lines aren't needed, but maybe a comment that end_key doesn't not need inf_suffix because we're looking at < end_key?

Or include start_has_inf_suffix = false for completeness?

spetrunia replied:
Removed the lines, added a comment.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+    else if (find_flag == HA_READ_AFTER_KEY)
+      start_has_inf_suffix= true;
+    else
+      DBUG_ASSERT(0);
```
If we hit this case, should this just set big_range = true?

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+
+  if (kd.m_is_reverse_cf) {
+    // Flip the endpoints
+    start_endp =rocksdb::Endpoint(end_slice, !end_has_inf_suffix);
```
It's not clear why the suffixes are inverted here?

If the forward cf range to be locked is (pk=10, 0) -> (pk=12, +inf), then isn't the reverse cf range to be locked
(pk=12, +inf) -> (pk=10, 0)? It seems (pk=12, 0) -> (pk=10, +inf) would not be right?

spetrunia replied:
The reason suffixes are inverted is as follows:

Let's consider a forward-ordered CF. The endpoint values would compare as follows (here LOWEST=0, HIGHEST=1, missing inf_suffix means any inf_suffix value):
 
```
 key=a, inf_suffix=LOWEST
 key=ab
 key=az
 key=a, inf_suffix=HIGHEST

 key=b, inf_suffix=LOWEST
 key=ba
 key=bz
 key=b, inf_suffix=HIGHEST
 ...
```
Putting the same set of keys into a reverse CF:

```
 key=b, inf_suffix=HIGHEST
 key=bz
 key=ba
 key=b, inf_suffix=LOWEST

 key=a, inf_suffix=HIGHEST
 key=az
 key=ab
 key=a, inf_suffix=LOWEST.
```
Note that the flag values have swapped their meanings, HIGHEST comes before LOWEST.

However, the endpoint comparison function in RocksDB doesn't "know" if the CF is reverse-ordered or not (it just makes use of the Key Comparator).

As for inf_suffix value, `Endpoint(key=$VAL, inf_suffix=0)` always compares before any endpoint with key=$VAL, and `Endpoint(key=$VAL, inf_suffix=1)` always compares after any endpoint with key=$VAL.

This is why one needs to flip the inf_suffix values.

spetrunia replied:
... added a comment about this.

hermanlee commented on storage/rocksdb/rdb_iterator.cc:
```diff
@@ -0,0 +1,974 @@
+
+  // Save the value of m_use_locking_iter because release_scan_iterator()
+  // will set it to false.
+  bool use_locking_iter= m_use_locking_iter;
```
It doesn't look like m_use_locking_iter is set back to use_locking_iter after release_scan_iterator() is called. 

In index_read_intern(), for read-committed, we could call setup_scan_iterator and release_iterator multiple times. If we ever did that for repeatable-read, then it seems m_use_locking_iter, then while the first couple of times through the loop might grab a locking iterator, the third and later times through the loop would not allocate a locking iterator.

spetrunia replied:
Addressed

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,190 @@
+        to check them here
+      */
+      DEBUG_SYNC(my_core::thd_get_current_thd(), "rocksdb.locking_iter_scan");
+      auto end_key = iter_->key();
```
Would it might make sense to store the range locked so far. If this key falls within the range, we won't have to perform another GetRangeLock(). Thinking ahead, if we wanted to optimize this by scan ahead a few keys and locking the whole range first, we can reduce the number of GetRangeLock() calls.

spetrunia replied:
> If this key falls within the range, we won't have to perform another GetRangeLock().

Are you talking about the case where another transaction is inserting records into the range we're scanning? Unless I am missing something, this is the only scenario where the current code will make more GetRangeLock() calls than necessary.

I agree that if we scan ahead several values, we'll need to remember what range we've locked.

hermanlee replied:
Yes, I'm thinking of the case where our snapshot gives us a next_key and as we attempt to acquire a range lock up to next_key, a concurrent transaction inserts and commits multiple keys before our next_key. Once we lock up to next_key, we refresh the snapshot and discover the next few committed keys is already in the range lock up to next_key.

spetrunia replied:
Ok, done.

spetrunia committed:
Range Locking: get tests to pass: select_count_for_update, skip_Locked_nowait.

spetrunia committed:
Address review input part #1

spetrunia committed:
Address review input part 2

spetrunia committed:
Address review input part #3

spetrunia committed:
Address review input part #4: add comment, fix coding style

spetrunia committed:
Range Locking: don't lock ranges when using READ COMMITTED isolation level

spetrunia committed:
Range Locking: make LockingIterator not lock the same range twice.

LockingIterator may attempt to lock the same range multiple times
if another transaction is inserting records into the range it is scanning
through. Make the code avoid this.

hermanlee reviewed (commented):
I'm seeing a different behavior from InnoDB where the gap from the last key to supremum is not being acquired.

On InnoDB:

TXN 1:
create table t1 (pk int primary key) engine=innodb;
insert into t1 values (1), (10), (100);
begin;
select * from t1 for update;

TXN 2:
begin;
select * from t1;
insert into t1 values (150); # times out on lock

On MyRocks:
insert into t1 values (150); # succeeds

hermanlee reviewed (commented):
Another difference with InnoDB:

```
TXN 1:
create table t1 (pk int primary key) engine=rocksdb;
insert into t1 values (1), (10), (100);
begin;
select * from t1;

TXN 2:
begin;
select * from t1;
insert into t1 values (90);
commit;

TXN 1:
select * from t1 lock in share mode; /* InnoDB returns 4 rows, but MyRocks only returns 3 rows: 90 is missing. */
```

Using `select * from t1 for update`, MyRocks correctly returns 4 rows instead of 3.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -4287,17 +4400,35 @@ class Rdb_transaction_impl : public Rdb_transaction {
+        Use InnoDB-like concurrency mode: make the DML statements always read
+        the latest data (instead of using transaction's snapshot).
+        This "downgrades" the transaction isolation to READ-COMMITTED on the
+        master, but in return the actions can be replayed on the slave.
```
For terminology, please use `primary` instead of `master` and `secondary` instead of `slave`.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -6667,6 +6882,15 @@ static int rocksdb_init_internal(void *const p) {
+  if (range_lock_mgr)
+  {
+    range_lock_mgr->SetMaxLockMemory(rocksdb_max_lock_memory);
+    sql_print_information("RocksDB: USING NEW RANGE LOCKING");
```
Can probably remove `NEW`. It's just range locking vs. point locking.

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,222 @@
+//////////////////////////////////////////////////////////////////////////////
+
+//
+// LockingIterator is an iterator that locks the rows before returning, as well
```
I don't think this actually works with non-materialized partial indexes. In the case where partial index is not materialized, the rows are read from the primary key, sorted in the secondary key order and then returned directly by the partial_index iterator. It would not make Seek()/Next() calls on the secondary index down to rocksdb, and so the range locks in the non-materialized partial index case would not be acquired through this iterator.

I'm getting an assertion when trying to test this though:

```
create table t1 (a int, b int, c int, d int, e int default 0  primary key(a, b, c), key (a, b, d) comment 'partial_group_keyparts=2,partial_group_threshold=10000') engine=rocksdb;
insert into t1 (a,b,c,d) values (1, 1, 100, 1000), (1,1,200, 2000), (1, 1, 300, 3000), (1, 1, 400, 4000);
update t1 set d = d + 1 where a = 1 and b = 1 and d > 2000; 
```

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7e124ed in __GI_abort () at abort.c:79
#2  0x00007ffff7e224d3 in __assert_fail_base (fmt=0x7ffff7f9a088 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x3eda1fc "!key_descr.is_primary_key() || (key_descr.is_primary_key() && m_row_lock_count > 0)", 
    file=file@entry=0x3cdbdc3 "/home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc", line=line@entry=4040, 
    function=function@entry=0x3a0cb52 "virtual void myrocks::Rdb_transaction_impl::release_lock(const myrocks::Rdb_key_def &, const std::string &, bool)") at assert.c:92
#3  0x00007ffff7e22543 in __GI___assert_fail (
    assertion=0x3eda1fc "!key_descr.is_primary_key() || (key_descr.is_primary_key() && m_row_lock_count > 0)", 
    file=0x3cdbdc3 "/home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc", line=4040, 
    function=0x3a0cb52 "virtual void myrocks::Rdb_transaction_impl::release_lock(const myrocks::Rdb_key_def &, const std::string &, bool)")
    at assert.c:101
#4  0x0000000007b54dfb in myrocks::Rdb_transaction_impl::release_lock (this=0x7fffbc84aa80, key_descr=..., 
    rowkey="\000\000\001\004\200\000\000\001\200\000\000\001\200\000", <incomplete sequence \310>, force=false)
    at /home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc:4039
#5  0x0000000007b0337c in myrocks::ha_rocksdb::unlock_row (this=0x7fffbc883428) at /home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc:10463
#6  0x0000000005cdd1a0 in Sql_cmd_update::update_single_table (this=0x7fffbc80faf0, thd=0x7fffbc80b000)
    at /home/herman/rocks-mysql/8.0/sql/sql_update.cc:809
#7  0x0000000005ce0c47 in Sql_cmd_update::execute_inner (this=0x7fffbc80faf0, thd=0x7fffbc80b000)
    at /home/herman/rocks-mysql/8.0/sql/sql_update.cc:1614
#8  0x0000000005bf58be in Sql_cmd_dml::execute (this=0x7fffbc80faf0, thd=0x7fffbc80b000) at /home/herman/rocks-mysql/8.0/sql/sql_select.cc:738
#9  0x0000000005b51f1e in mysql_execute_command (thd=0x7fffbc80b000, first_level=true, last_timer=0x7ffff53b39b0)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:4204
#10 0x0000000005b4b38d in mysql_parse (thd=0x7fffbc80b000, parser_state=0x7ffff53b4560, last_timer=0x7ffff53b39b0)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:6202
#11 0x0000000005b46cf9 in dispatch_command (thd=0x7fffbc80b000, com_data=0x7ffff53b4d98, command=COM_QUERY)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:2226
#12 0x0000000005b49ee1 in do_command (thd=0x7fffbc80b000) at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:1560
#13 0x0000000005dcaa21 in handle_connection (arg=0x7fffcea982e0)
    at /home/herman/rocks-mysql/8.0/sql/conn_handler/connection_handler_per_thread.cc:330
#14 0x00000000079f00fb in pfs_spawn_thread (arg=0x7ffff748a620) at /home/herman/rocks-mysql/8.0/storage/perfschema/pfs.cc:2884
#15 0x00007ffff7d5d20c in start_thread (arg=0x7ffff53b6700) at pthread_create.c:479
#16 0x00007ffff7f0581f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

However, performance with partial indexes might be very poor because we sort the primary keys, set the range lock, and then we need to re-read/re-sort the primary keys. We'd want to lock based on the rows returned by the partial index iterator.

Alternatively, we could skip supporting range locks for partial indexes and use only point locks, but going down this road means the compatibility matrix among different myrocks features like ttl, range locking, partial indexes, etc. gets complicated quickly.

cc: @lth

hermanlee replied:
This assertion happens even with a regular SK (not partial index) in this table schema.

hermanlee replied:
Some further thoughts. I think we can either range lock the PK or range lock the SK here. Range locking the PK would still be correct since any insert/updates/scans would still need to lock the PK. We may end up locking more than needed, but I think it simplifies the implementation.

Since the PK and SK share the same prefixes and the PK is read to find the prefixes needed for the SK, we can range lock the prefixes on the PK in get_prefix_from_start(). m_pk_iterator.seek() is used to find the next prefixes, so we need to range lock the start_key to the returned prefix. We can range lock from endpoint (start_key, 0) to the endpoint (returned_prefix, +inf) so that we don't need to individually lock each gap between rows of the returned_prefix.

We would need to also implement the read/lock/re-read logic using the m_pk_iterator here. Given how the partial index iterator uses m_pk_iterator, it's not clear to me pushing m_iter_should_use_locking to the m_pk_iterator is sufficient because the range locks need to cover the whole range from one prefix to the next.

lth replied:
Currently, partial indexes will call always Seek() on the secondary key range before reading from the PK, and that theoretically should acquire the correct range locks, and partial indexes should just work without much additional changes.

What we'll end up is that the locking PK iterator will ensure that no new prefixes show up in between consecutive prefixes, and the locking SK iterator will ensure that no new keys can be added within a prefix. 

That said, I think there are a few problems:
1. I don't think we're acquiring range locks if the iterator is invalid (in general, it doesn't seem like we acquire range locks up to the iterator end bound).
2. We also need to make sure we're acquiring point locks on the SK even if they're not materialized (it doesn't seem like that is done in this diff though).
3. In the partial index seek code, we always expect that we'd be able to read some rows out after getting the next prefix from the PK. However, with iterator refreshing, that might not be true anymore, so we probably need to add a while loop there.

There also may be some opportunity for optimization, since partial indexes currently uses a point lock on the prefix key to stand in as a range lock over the prefix (we do something similar for unique keys), so we probably don't have to lock twice. Also, it might be cleaner to lock only on PK instead as described above.

hermanlee replied:
Yes, Rdb_iterator_partial::seek() does always call Rdb_iterator_base::seek() to lock the range. We should get the correct range locking out of the partial index iterator's base iterator. The main downside is the ranges might be very large if the SK doesn't have any materialized prefixes, but that's a performance/concurrency issue, and not a correctness one.

I'd think we'd also need to set m_iterator_pk.m_iter_should_use_locking based on the partial index iterator's setting, which I think is not currently happening. And yes, I think locking only on the PK might be cleaner. We might be able to get away with not having to acquire SK point locks if we lock the whole prefix range in the PK before doing anything.

spetrunia replied:
> This assertion happens even with a regular SK (not partial index) in this table schema.

Confirm.  I've checked the cause:  The code attempts to count `m_row_lock_count`, and it counts range lock acquisition as one lock.  For rows that didn't match the WHERE, `unlock_row` decrements the count. If there is more than one row in the range, the counter can go below zero and fire an assert.

spetrunia replied:
Ok finally committed 9f29d708d47db4aaf4e015697dd088d2ee408b44  (pushed into this PR, too)  with a solution for this.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -10850,6 +11331,15 @@ int ha_rocksdb::update_write_sk(const TABLE *const table_arg,
         reinterpret_cast<const char *>(m_sk_packed_tuple_old), old_packed_size);
 
+    /* Range locking: lock the index tuple being deleted */
+    if (m_use_range_locking) {
```
It seems like this lock should go into check_and_lock_sk() instead?

hermanlee replied:
Also, it seems like this check should be using  `rocksdb_use_range_locking` instead? If the range lock manager is being used, then a singlepoint lock is needed on each secondary key regardless of the current transaction's isolation level.

For example:

```
TXN 1:

create table i1 (pk int, a int, b int, primary key (pk), key(a)) engine=innodb;
insert into i1 values (1, 100, 1000), (2, 200, 2000), (3, 300, 3000);
set transaction isolation level repeatable read;
begin;
update i1 set b = b + 1 where a > 200;

TXN 2:

set transaction isolation level read committed;
begin;
insert into i1 values (4, 150, 1500); /* Success */
insert into i1 values (5, 250, 1500); /* Blocks on InnoDB, but succeeds on MyRocks. */
```

spetrunia replied:
> Also, it seems like this check should be using rocksdb_use_range_locking instead?


Agree. done.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -11900,8 +12411,15 @@ int ha_rocksdb::external_lock(THD *const thd, int lock_type) {
     }
     tx->m_n_mysql_tables_in_use++;
-    rocksdb_register_tx(rocksdb_hton, thd, tx);
+    rocksdb_register_tx(rocksdb_hton, thd, tx, (lock_type == F_WRLCK));
```
This seems like it should be `lock_type == F_WRLCK || lock_type == F_RDLCK` so that `select lock in share mode` can also read the latest data?

spetrunia replied:
Agree that `select ... lock in share mode` should read the latest data.
However, `F_RDLCK` is passed for both SELECTs and `SELECT ... LOCK IN SHARE MODE`. 
Changed the check to analyze `m_lock_rows`.

hermanlee commented on storage/rocksdb/rdb_i_s.cc:
```diff
@@ -1771,31 +1771,63 @@ static int rdb_i_s_lock_info_fill_table(
-
-      if (ret != 0) {
-        break;
+  if (range_lock_mgr) {
```
Given how similar the range lock and non-range lock sections are, would it be better to refactor this code? The code populating the tables->table->field could be added to separate helper function that takes a std::string for the key_hexstr and char for lock_mode.

spetrunia replied:
Done

hermanlee commented on storage/rocksdb/rdb_locking_iter.h:
```diff
@@ -0,0 +1,222 @@
+          continue;
+        }
+      } else {
+        // There's no row (within the iterator bounds perhaps). Exit now.
```
It seems like we may need range lock between the target slice and the end range of the iterator bounds.

spetrunia replied:
Yes, done.

spetrunia replied:
Wait, the fix fix crashes with `rocksdb_enable_iterate_bounds=off` . Will fix

spetrunia replied:
Now done.

spetrunia committed:
Address review input part #1

spetrunia committed:
Address review input: lock the "gap" after the last row that was read

spetrunia committed:
With range locking, m_row_lock_count is incremented by 1 for each range

Do not decrement it in unlock_row, as there may be multiple unlock_row
calls per range.

spetrunia committed:
Update rocksdb.trx_info test

Range Locking code locks the gap after the last value now, which
means for 2-row table "SELECT * FROM t FOR UPDATE" will place 3 locks.

spetrunia committed:
Initialize ha_rocksdb::m_use_range_locking.

Initializing it in ha_rocksdb::external_lock was not enough. MyRocks uses
index read calls before external_lock() is called:

myrocks::Rdb_iterator_base::seek()
myrocks::ha_rocksdb::index_read_intern()
myrocks::ha_rocksdb::index_read_intern()
myrocks::ha_rocksdb::index_last()
myrocks::ha_rocksdb::load_auto_incr_value_from_index()
myrocks::ha_rocksdb::load_auto_incr_value()
myrocks::ha_rocksdb::open()
handler::ha_open()

spetrunia committed:
Make LockingIterator work when rocksdb_enable_iterate_bounds=off

This is a followup to:

  Address review input: lock the "gap" after the last row that was read

Made the code work when iterator bounds are not available, in which
case it locks till index start/end.

spetrunia committed:
Fix a typo

spetrunia committed:
Range Locking: READ COMMITTED doesnt take range locks but should still lock secondary keys

spetrunia committed:
Code cleanup: Make LockingIterator's methods accept direction as parameter

.. not as template argument.

spetrunia committed:
Range Locking: SELECT ... LOCK IN SHARE MODE should skip snapshot, too.

spetrunia committed:
Range Locking: Move locking secondary keys to right place

From ha_rocksdb::update_write_sk into ha_rocksdb::check_and_lock_sk

spetrunia committed:
Range Locking: Make partial index support LockingIterator.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -4346,17 +4460,35 @@ class Rdb_transaction_impl : public Rdb_transaction {
+        Use InnoDB-like concurrency mode: make the DML statements always read
+        the latest data (instead of using transaction's snapshot).
+        This "downgrades" the transaction isolation to READ-COMMITTED on the
+        primary, but in return the actions can be replayed on the replica.
```
The first half of the sentence makes sense, but the second half: "but in return the actions can be replayed on the replica" seems to involve replication and it's not clear how replication to secondaries is affected by this concurrency mode?

spetrunia replied:
Agree, it doesn't sound clear. 

IIRC I had this in mind when writing this: 

With statement-based replication and point locking, one can get the slave 
out of sync. Using gap(or range) locking allows to avoid that. But when 
one uses Gap (or range) locking, the locking reads must read the latest 
committed data.

I guess this should either be stated in full, or not at all...

hermanlee commented on storage/rocksdb/rdb_locking_iter.cc:
```diff
@@ -0,0 +1,286 @@
+        continue;
+      }
+    } else {
+      // Some error
```
Seems like this comment is inaccurate? If we reached the end of the range, it's not really an error?

spetrunia replied:
Agree... "Some error" here included EoF which is not really an error.

hermanlee reviewed (commented):
The following doesn't look right:

```
/* TXN 1 */
mysql> create table t1 (pk int not null auto_increment, a int, b int, primary key (pk), unique key (a)) engine=rocksdb;
Query OK, 0 rows affected (0.05 sec)
                                         
mysql> begin;
Query OK, 0 rows affected (0.00 sec)
                                                                                   
mysql> insert into t1 (a, b) values (1, 1), (2, 2);                               
Query OK, 2 rows affected (0.01 sec)                                               
Records: 2  Duplicates: 0  Warnings: 0                                             
                                                                                   
mysql> select * from information_schema.rocksdb_locks;                            
+------------------+----------------+--------------------+------+
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                | MODE |
+------------------+----------------+--------------------+------+
|                0 |             17 | 000001080180000002 | X    |
|                0 |             17 | 000001080180000001 | X    |
|                0 |             17 | 0000010780000002   | X    |
|                0 |             17 | 0000010780000001   | X    |
+------------------+----------------+--------------------+------+
4 rows in set (0.01 sec)

/* TXN 2 */
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 (a, b) values (10, 10), (20, 20);
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction: Timeout on index: test.t1.PRIMARY

/* TXN 1 */
mysql> select * from information_schema.rocksdb_locks;
+------------------+----------------+-------------------------------------+------+
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                                 | MODE |
+------------------+----------------+-------------------------------------+------+
|                0 |             17 | 0000010780000001-000001080180000002 | X    |
+------------------+----------------+-------------------------------------+------+
1 row in set (0.01 sec)
```

The range lock seems to span multiple indexes.

Another example:

```
mysql> create table t6 (pk int not null auto_increment, a int, b int, c int, primary key (pk), unique key (a), key (b), unique key (c)) engine=rocksdb;
Query OK, 0 rows affected (0.03 sec)
                                          
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t6 (a, b, c) values (1, 1, 1), (2, 2, 2);                                                                                                             
Query OK, 2 rows affected (0.00 sec)                                                                                                                                     
Records: 2  Duplicates: 0  Warnings: 0    
                                          
mysql> select * from information_schema.rocksdb_locks;                                                                                                                   
+------------------+----------------+---------------------------------------------+------+                                                                               
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                                         | MODE |                                                                               
+------------------+----------------+---------------------------------------------+------+                                                                               
|                0 |             21 | 00000113018000000280000002                  | X    |                                                                               
|                0 |             21 | 0000011080000001-00000113018000000180000001 | X    |                                                                               
+------------------+----------------+---------------------------------------------+------+                                                                               
2 rows in set (0.00 sec)                  
                                                                                                                                                                         
mysql> select index_name, hex(index_number) from information_schema.rocksdb_ddl where table_name = 't6';                                                                 
+------------+-------------------+
| index_name | hex(index_number) |  
+------------+-------------------+        
| PRIMARY    | 110               |
| a          | 111               |  
| c          | 112               |        
| b          | 113               |                                                                                                                                       
+------------+-------------------+  
4 rows in set (0.00 sec)

spetrunia commented:
https://github.com/facebook/mysql-5.6/pull/1165#pullrequestreview-769443923
> The following doesn't look right: 
> ...

I can't reproduce this behavior... This is what I'm getting:

```
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 (a, b) values (1, 1), (2, 2);
Query OK, 2 rows affected (0.01 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select * from information_schema.rocksdb_locks;
+------------------+----------------+--------------------+------+
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                | MODE |
+------------------+----------------+--------------------+------+
|                0 |              2 | 0000016e0180000002 | X    |
|                0 |              2 | 0000016d80000002   | X    |
|                0 |              2 | 0000016e0180000001 | X    |
|                0 |              2 | 0000016d80000001   | X    |
+------------------+----------------+--------------------+------+
4 rows in set (0.00 sec)
```
so far the same. But TXN 2 will succeed to get the lock:

```
/* TXN 2 */

mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 (a, b) values (10, 10), (20, 20);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select * from information_schema.rocksdb_locks;
+------------------+----------------+--------------------+------+
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                | MODE |
+------------------+----------------+--------------------+------+
|                0 |              3 | 0000016e0180000014 | X    |
|                0 |              3 | 0000016e018000000a | X    |
|                0 |              2 | 0000016e0180000002 | X    |
|                0 |              2 | 0000016e0180000001 | X    |
|                0 |              3 | 0000016d80000004   | X    |
|                0 |              3 | 0000016d80000003   | X    |
|                0 |              2 | 0000016d80000002   | X    |
|                0 |              2 | 0000016d80000001   | X    |
+------------------+----------------+--------------------+------+
```
..which looks correct to me

spetrunia commented:
> Another example:
> 
> ```
> mysql> create table t6 (pk int not null auto_increment, a int, b int, c int, primary key (pk), unique key (a), key (b), unique key (c)) engine=rocksdb;
> Query OK, 0 rows affected (0.03 sec)
>                                           
> mysql> begin;
> Query OK, 0 rows affected (0.00 sec)
> 
> mysql> insert into t6 (a, b, c) values (1, 1, 1), (2, 2, 2);                                                                                                             
> Query OK, 2 rows affected (0.00 sec)                                                                                                                                     
> Records: 2  Duplicates: 0  Warnings: 0    
>                                           
> mysql> select * from information_schema.rocksdb_locks;                                                                                                                   
> +------------------+----------------+---------------------------------------------+------+                                                                               
> | COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                                         | MODE |                                                                               
> +------------------+----------------+---------------------------------------------+------+                                                                               
> |                0 |             21 | 00000113018000000280000002                  | X    |                                                                               
> |                0 |             21 | 0000011080000001-00000113018000000180000001 | X    |                                                                               
> +------------------+----------------+---------------------------------------------+------+                                                                               
> 2 rows in set (0.00 sec)                  
> ```

For me, I get this:
```
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t6 (a, b, c) values (1, 1, 1), (2, 2, 2);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> 
mysql> select * from information_schema.rocksdb_locks;
+------------------+----------------+----------------------------+------+
| COLUMN_FAMILY_ID | TRANSACTION_ID | KEY                        | MODE |
+------------------+----------------+----------------------------+------+
|                0 |              1 | 00000172018000000280000002 | X    |
|                0 |              1 | 000001710180000002         | X    |
|                0 |              1 | 000001700180000002         | X    |
|                0 |              1 | 0000016f80000002           | X    |
|                0 |              1 | 00000172018000000180000001 | X    |
|                0 |              1 | 000001710180000001         | X    |
|                0 |              1 | 000001700180000001         | X    |
|                0 |              1 | 0000016f80000001           | X    |
+------------------+----------------+----------------------------+------+
8 rows in set (0.00 sec)

mysql>  select index_name, hex(index_number) from information_schema.rocksdb_ddl where table_name = 't6';
+------------+-------------------+
| index_name | hex(index_number) |
+------------+-------------------+
| PRIMARY    | 16F               |
| a          | 170               |
| c          | 171               |
| b          | 172               |
+------------+-------------------+
4 rows in set (0.01 sec)
```
which looks ok at first glance

spetrunia commented:
Double checking if I am using the right tree... it is https://github.com/spetrunia/mysql-5.6/commits/fb-mysql-8.0.20-range-locking-may 
```
commit 9f29d708d47db4aaf4e015697dd088d2ee408b44 (HEAD -> fb-mysql-8.0.20-range-locking-may, origin/fb-mysql-8.0.20-range-locking-may)
Author: Sergei Petrunia <psergey@askmonty.org>
Date:   Mon Sep 13 20:23:12 2021 +0300

    Range Locking: Make partial index support LockingIterator.

```

spetrunia committed:
Remove out-of-date comment.

hermanlee commented:
I ran my tests with the range locking change based on top of: https://github.com/facebook/mysql-5.6/commit/b22013dba3a9f0d242004f71bcac4427c7249abb.

In your lock outputs, none of the locks look like they were escalated and merged together. What triggers that process?

spetrunia commented:
lock escalation is triggered by lock tree memory exceeding the limit.  The current lock memory usage should be shown in `rocksdb_current_lock_memory`, and  `rocksdb_max_lock_memory` is where one sets the limit.

hermanlee commented:
Ah, my setup also inherited these settings from an existing test:

-rocksdb_use_range_locking=1 --rocksdb_max_lock_memory=1024

spetrunia committed:
An attempt at concurrent test for Range Locking

spetrunia committed:
Small fixes in range_locking_conc_test

- Truncate the table before starting the test.
- Added n_insert_failed counter

spetrunia committed:
range_locking_conc_test: if we found and deleted a group, create a new one

spetrunia committed:
Update range_locking_conc_test.test

spetrunia committed:
range_locking_conc_test: Do the requested number of operations

If an attempt to insert a row or create a group fails (because we've
found unsuitable data), do not count it.

spetrunia committed:
Code cleanup

spetrunia committed:
range_locking_conc_test.py: group creation shouldn't just rollback

It starts by doing a locking read of 5 rows in a random location.

If it encounters a grouped row that's not a group leader, it should
follow till the end of this group and then delete+create the next group.

spetrunia committed:
range_locking_conc_test: fixes to achieve greater contention

Also: We are now using skewed distribution, in order for INSERTs
to not become stuck, make them do deletes when they are not able
to insert.

laurynas-biveinis commented:
Should this one be closed because of https://github.com/facebook/mysql-5.6/pull/1185 ?
